File: sdks/python/examples/02_upload_image.py
#!/usr/bin/env python3
"""02 · Upload an image — conditioning a model on a file.

`upload_file` returns a `FileRef`; pass it into a command as a value and the
model receives the file. Helios takes prompt and image together through
`set_conditioning` so `start` cannot observe a half-set session.

    uv run python examples/02_upload_image.py ref.png

Docs: https://docs.reactor.inc/concepts/file-uploads
      https://docs.reactor.inc/model-api-reference/helios/schema
"""

from __future__ import annotations

import asyncio
import os
import sys

import display

from reactor_sdk import DEFAULT_API_URL, Reactor

API_KEY = os.environ.get("REACTOR_API_KEY")
MODEL = os.environ.get("REACTOR_MODEL", "helios")
SECONDS = float(os.environ.get("REACTOR_SECONDS", "15"))
SHOW = os.environ.get("REACTOR_SHOW") == "1"

PROMPT = "the same scene at night, lit by a campfire"
OUTPUT_TRACK = "main_video"


async def main() -> None:
    if len(sys.argv) != 2:
        sys.exit(f"usage: {sys.argv[0]} <image>")
    image = sys.argv[1]
    if not API_KEY and os.environ.get("REACTOR_LOCAL") != "1":
        sys.exit("set REACTOR_API_KEY — https://www.reactor.inc/account/api-keys")

    reactor = Reactor(
        model_name=MODEL,
        api_key=API_KEY,
        api_url=os.environ.get("REACTOR_API_URL", DEFAULT_API_URL),
        local=os.environ.get("REACTOR_LOCAL") == "1",
    )

    @reactor.on_status
    def status(new: str) -> None:
        print(f"status: {new}")

    @reactor.on_error
    def failed(error: Exception) -> None:
        print(f"error: {error}")

    # A refused upload arrives as `command_error`, not as a failed call.
    @reactor.on_message
    def message(msg: dict) -> None:
        print(f"message: {msg}")

    async with reactor:
        await reactor.connect()
        print(f"session: {reactor.session_id}")

        # Needs a ready session: the bytes go to that session's store. Name and
        # MIME type are inferred from the path.
        uploaded = await reactor.upload_file(image)
        print(f"uploaded: {uploaded.name} {uploaded.mime_type} ({uploaded.size} bytes)")

        await reactor.send_command("set_conditioning", {"prompt": PROMPT, "image": uploaded})
        await reactor.send_command("start", {})

        output = reactor.track(OUTPUT_TRACK)
        frames = 0
        window = display.window(f"{MODEL} · {uploaded.name}", enabled=SHOW)

        @output.on_raw_frame
        def count(data: bytes, width: int, height: int, *_: object) -> None:
            nonlocal frames
            frames += 1
            window.submit(data, width, height)

        await window.hold(SECONDS)
        print(f"frames: {frames}")


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("interrupted")

File: sdks/cpp/examples/02_upload_image.cpp
// 02 — Upload a file, then pass the reference into a command.
//
// The bytes cross the wire once: `upload_file` hands them to the platform and
// returns a `FileRef`, and the command carries the *reference*. Helios takes a
// prompt and an image together through `set_conditioning`, so the two arrive as
// one atomic change rather than as two the model has to reconcile.
//
//   export REACTOR_API_KEY=...
//   ./02_upload_image path/to/image.png
//
// Docs: https://docs.reactor.inc/concepts/file-uploads

#include <atomic>
#include <chrono>
#include <cstdlib>
#include <iostream>
#include <reactor/reactor.hpp>
#include <string>
#include <thread>

#include "display.hpp"

namespace {

constexpr auto MODEL = "reactor/helios";
constexpr auto PROMPT = "the same scene at golden hour, cinematic";

}  // namespace

int main(int argc, char** argv) {
  const char* api_key = std::getenv("REACTOR_API_KEY");
  if (api_key == nullptr) {
    std::cerr << "set REACTOR_API_KEY\n";
    return 2;
  }
  if (argc != 2) {
    std::cerr << "usage: " << argv[0] << " <image>\n";
    return 2;
  }
  const std::string image = argv[1];

  reactor::Reactor client{MODEL, reactor::ApiKey{api_key}};

  // A refused upload arrives as `command_error` from the model, not as a failed
  // call: the platform accepted the file, and the model declined to use it.
  auto messages = client.on_message([](const reactor::Json& message) {
    if (message.value("type", std::string{}) == "command_error") {
      std::cerr << "the model refused a command: " << message.dump() << '\n';
    }
  });
  auto errors = client.on_error(
      [](const reactor::ReactorError& error) { std::cerr << "error: " << error.what() << '\n'; });

  try {
    client.connect().get();
    std::cout << "session " << client.session_id().value_or("?") << " is ready\n";

    const auto uploaded = client.upload_file(image).get();
    std::cout << "uploaded: " << uploaded.name << ' ' << uploaded.mime_type << " (" << uploaded.size
              << " bytes)\n";

    // The reference goes in as a named upload, not as an argument: the platform
    // resolves it on its side.
    client.send_command("set_conditioning", {{"prompt", PROMPT}}, {{"image", uploaded}}).get();
    client.send_command("start").get();

    std::atomic<int> frames{0};
    examples::Display display{std::string{"02 — "} + uploaded.name};
    auto video = client.track("main_video");
    auto subscription = video.on_frame([&](const reactor::VideoFrame& frame) {
      ++frames;
      display.show(frame);
    });

    const auto deadline = std::chrono::steady_clock::now() + std::chrono::seconds(10);
    while (std::chrono::steady_clock::now() < deadline && display.pump()) {
      std::this_thread::sleep_for(std::chrono::milliseconds(20));
    }

    std::cout << "received " << frames.load() << " frames conditioned on " << uploaded.name << '\n';
    client.disconnect().get();
    return frames.load() > 0 ? 0 : 1;
  } catch (const reactor::ReactorError& error) {
    std::cerr << "failed: " << error.what() << '\n';
    try {
      client.disconnect().get();
    } catch (const reactor::ReactorError&) {
    }
    return 1;
  }
}
